hibernate-core/src/main/java/org/hibernate/dialect/function/SybaseTruncFunction.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.dialect.function;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.hibernate.query.ReturnableType;
+import org.hibernate.query.spi.QueryEngine;
+import org.hibernate.query.sqm.function.SelfRenderingSqmFunction;
+import org.hibernate.query.sqm.tree.SqmTypedNode;
+import org.hibernate.query.sqm.tree.expression.SqmExtractUnit;
+import org.hibernate.type.spi.TypeConfiguration;
+
+/**
+ * Custom {@link TruncFunction} for Sybase which uses the dialect-specific emulation function for datetimes
+ *
+ * @author Marco Belladelli
+ */
+public class SybaseTruncFunction extends TruncFunction {
+	public SybaseTruncFunction() {
+		super(
+				"sign(?1)*floor(abs(?1))",
+				"sign(?1)*floor(abs(?1)*power(10,?2))/power(10,?2)",
+				null,
+				null
+		);
+	}
+
+	@Override
+	protected <T> SelfRenderingSqmFunction<T> generateSqmFunctionExpression(
+			List<? extends SqmTypedNode<?>> arguments,
+			ReturnableType<T> impliedResultType,
+			QueryEngine queryEngine,
+			TypeConfiguration typeConfiguration) {
+		final List<SqmTypedNode<?>> args = new ArrayList<>( arguments );
+		if ( arguments.size() == 2 && arguments.get( 1 ) instanceof SqmExtractUnit ) {
+			// datetime truncation
+			return new SybaseDateTruncEmulation( "convert", typeConfiguration ).generateSqmFunctionExpression(
```

Comment:
Would be nice if you could keep the `SybaseDateTruncEmulation` in a private final variable initialized in the constructor.

Reply:
Should I also make dialect-specific datetime trunc emulation functions inner classes of the already dialect-specific trunc functions (e.g. `SybaseTruncFunction$SybaseDateTruncEmulation`)?

Reply:
Don't know if I understand exactly, but whatever you think works best ;)